src/lib/concurrency/pipe_lib/test/test_pipe_lib.ml

```diff
@@ -0,0 +1,147 @@
+      if !counter = expected_length then Option.iter ~f:kill pipe ;
+      return () )
+  >>| fun () ->
+  if !counter = expected_length then ()
+  else
+    failwithf "unexpected number of elements: expected %d, got %d"
+      expected_length !counter ()
+
+let create_buffered_swappable capacity =
+  Pipe_lib.Strict_pipe.Swappable.create ~name:"swappable"
+    (Buffered (`Capacity capacity, `Overflow (Drop_head (const ()))))
+
+let read_all_values_or_timeout ?pipe ~expected iterator =
+  let read_all_values = read_all_values iterator ~expected ?pipe in
+  Deferred.choose
+    [ Deferred.choice read_all_values ident
+    ; Deferred.choice
+        (after (Time_ns.Span.of_sec 1.5))
+        (fun () -> failwith "Swappable strict pipe hangs, timeout!")
+    ]
+
+let test_reader_hang () =
+  let open Pipe_lib.Strict_pipe.Swappable in
+  let swappable = create_buffered_swappable 5 in
+  let%bind _hanging_reader = swap_reader swappable in
+  write swappable 1 ;
+  write swappable 2 ;
+  let%bind good_reader = swap_reader swappable in
+  write swappable 3 ;
+  read_all_values_or_timeout good_reader ~pipe:swappable ~expected:[ 1; 2; 3 ]
+
+let test_buffered_overflow () =
+  let open Pipe_lib.Strict_pipe.Swappable in
+  let pipe = create_buffered_swappable 4 in
+  (* Write 7 elements to a pipe with capacity 5, first 2 should be dropped due to overflow *)
+  List.iter [ 1; 2; 3; 4; 5; 6; 7 ] ~f:(write pipe) ;
+  let%bind reader = swap_reader pipe in
+  (* It's unintuitive that 5 elements are returned when capacity is 4,
+     but this is the behavior of strict pipe, i.e. there seems to be an
+     off-by-one bug there. *)
+  read_all_values_or_timeout ~pipe ~expected:[ 3; 4; 5; 6; 7 ] reader
+
+let test_multiple_iterations () =
+  let open Pipe_lib.Strict_pipe.Swappable in
+  let pipe = create_buffered_swappable 10 in
+  let elements = [ 1; 2; 3; 4; 5 ] in
+  (* Write elements to the pipe *)
+  List.iter elements ~f:(write pipe) ;
+  (* Get a reader and verify we can iterate twice *)
+  let%bind reader = swap_reader pipe in
+  (* First iteration *)
+  let%bind () = read_all_values_or_timeout ~pipe ~expected:elements reader in
+  (* Second iteration - should see the same elements again *)
+  read_all_values_or_timeout ~expected:elements reader
+
+let test_concurrent_iterators () =
+  let open Pipe_lib.Strict_pipe.Swappable in
+  let pipe = create_buffered_swappable 10 in
+
+  (* Write initial two elements *)
+  List.iter ~f:(write pipe) [ 1; 2 ] ;
+
+  let%bind reader1 = swap_reader pipe in
+  (* Run first iteration *)
+  let%bind reader2 =
+    let first_iter_count = ref 0 in
+    let next_reader_ref = ref None in
+    let%map () =
+      Iterator.iter reader1 ~f:(fun _ ->
+          incr first_iter_count ;
+          if !first_iter_count = 2 then
+            (* After reading two elements, create second iterator and write more elements *)
+            let%map next_reader = swap_reader pipe in
+            next_reader_ref := Some next_reader
```

Comment:
A clever use, but I would avoid that in general. Is there any usecase you have on mind for this?